include/git2/repository.h

```diff
@@ -770,6 +770,31 @@ GIT_EXTERN(int) git_repository_ident(const char **name, const char **email, cons
  */
 GIT_EXTERN(int) git_repository_set_ident(git_repository *repo, const char *name, const char *email);
 
+/**
+ * Load and cache all submodules.
+ *
+ * Because the `.gitmodules` file is unstructured, loading submodules is an
+ * O(N) operation.  Any operation (such as `git_rebase_init`) that requires
+ * accessing all submodules is O(N^2) in the number of submodules, if it
+ * has to look each one up individually.  This function loads all submodules
+ * and caches them so that subsequent calls to `git_submodule_lookup` are O(1).
+ *
+ * @param repo the repository whose submodules will be cached.
+ */
+GIT_EXTERN(int) git_repository_submodule_cache_all(
```

Comment:
There's no indication of when somebody should call `cache_clear`.  You need to describe what operations would cause a cache coherency problem; it's not at all obvious to the reader and I suspect it's wildly dangerous to call `cache_all` and then (say) check out a new branch that would update a submodule?  I'm just guessing, I don't know, and I'd like this to expand upon that.

I also think that you should move this into `include/git2/sys/repository.h`.  This seems to me to be a cache that you have to manage rather carefully, and the `sys/` directory is for the things that will hurt you badly if you get wrong.